docker/bookkeeper/entrypoint.sh

```diff
@@ -48,31 +56,143 @@ export BK_tlsTrustStoreType=JKS
+    if [ $? -eq 0 ]; then
+        echo "Cluster metadata already exists"
+    else
+        # Create an ephemeral zk node `bkInitLock` for use as a lock.
+        lock=`zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH}/bkInitLock '' true false false" ${BK_zkServers}`
+        if [ -z "$lock" ]; then
+            echo "Bookkeeper znodes do not exist in Zookeeper. Initializing a new Bookeekeper cluster."
+
+            # Note that this `bookkeeper` shell script sets "exit on error" (`set -e`) and sources from
+            # ${SCRIPTS_DIR}/common.sh. Make sure to invoke `fix_bk_ipv6_check()` before the control reaches here.
+            /opt/bookkeeper/bin/bookkeeper shell initnewcluster
+            if [ $? -eq 0 ]; then
+                echo "initnewcluster operation succeeded"
+            else
+                echo "initnewcluster operation failed. Please check the reason."
+                echo "Exit status of initnewcluster"
+                echo $?
+                exit
+            fi
+        else
+            echo "Others may be initializing the cluster at the same time."
+            tenSeconds=1
+            while [ ${tenSeconds} -lt 100 ]
+            do
+                sleep 10
+                zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+                if [ $? -eq 0 ]; then
+                    echo "Waited $tenSeconds * 10 seconds. Successfully listed ''${BK_zkLedgersRootPath}/available/readonly'"
+                    break
+                else
+                    echo "Waited $tenSeconds * 10 seconds. Continue waiting."
+                    (( tenSeconds++ ))
+                    continue
+                fi
+            done
+
+            if [ ${tenSeconds} -eq 100 ]; then
+                echo "Waited 100 seconds for bookkeeper cluster to initialize, but to no avail. Something is wrong, please check."
+                exit
+            fi
+        fi
+    fi
+    set -e
+}
+
+format_bookie_data_and_metadata() {
+    if [ `find $BK_journalDirectory $BK_ledgerDirectories $BK_indexDirectories -type f 2> /dev/null | wc -l` -gt 0 ]; then
+      # The container already contains data in BK directories. This is probably because
+      # the container has been restarted; or, if running on Kubernetes, it has probably been
+      # updated or evacuated without losing its persistent volumes.
+      echo "Data available in bookkeeper directories; not formatting the bookie"
+    else
+      # The container does not contain any BK data, it is probably a new
+      # bookie. We will format any pre-existent data and metadata before starting
+      # the bookie to avoid potential conflicts.
+      echo "Formatting bookie data and metadata"
+      /opt/bookkeeper/bin/bookkeeper shell bookieformat -nonInteractive -force -deleteCookie || true
+    fi
+}
+
+echo "Creating directories for Bookkeeper journal and ledgers"
```

Comment:
In BookKeeper, we also have the option of setting the path for index directories. Is this a feature we might want to use in production, writing the ledger indexes to a separate directory/drive? If so, we need to provision for it.

Reply:
I am not sure about this, but I suppose it might be useful to support locating ledger indexes to a specified directory. However, I want to scope of this PR limited to upgrade to Bookkeeper `4.9.2`. This is also one of the preexisting logic. Hope that is OK.

Reply:
The reason to create customized journal and ledger files here is to support multi ledger/journal for better write performance as we find that by increasing the parallasim of writing to thoses files, the performance will be increased. However, during the experiment I find that index dir have a very low write throughput so it shouldn't have any negative effect to the performance by just using the default settings.

Reply:
Thanks @Tristan1900. @fpj I've added comments to explain the context in the code. Hope that helps.